src/main/java/se/liu/ida/hefquin/engine/data/VocabularyMapping.java

```diff
@@ -0,0 +1,25 @@
+package se.liu.ida.hefquin.engine.data;
+
+import java.util.Set;
+import org.apache.jena.reasoner.TriplePattern;
+
+
+public interface VocabularyMapping {
+
+	/**
+	 * Applies this vocabulary mapping to the given triple pattern and
+	 * returns the resulting triple patterns. If this mapping is not relevant
+	 * for the given triple pattern (i.e., applying the mapping to the given
+	 * triple pattern does not have any effect), then the result of this
+	 * function is a singleton set that simply contains the given triple
+	 * pattern.
+	 */
+	
+	/**
+	 * TODO: Is TripplePattern enough? Which data structure to use?
+	 * Unions and intersections also need to be represented
+	 * Would Graph be a fitting? Both Jena union and intersection are graphs
+	 */
```

Comment:
I assume that this question is about the return type and not the argument of this function.

If so, when I proposed this return type to be a set of triple patterns, I was assuming that the physical operator(s) that obtain this set will consider the triple patterns in this set to form a union. In other words, the physical operator may take these resulting triple patterns and issue a request for each of them, one after another; and, then, the solution mappings produced from the responses to these requests can all be output of the operator (which is, essentially, making a union of the responses).

However, now you mention intersection. I am not entirely sure where intersection may come into play. May understanding so far was that your approach uses a sequence of three functions (translateSubject, translatePredicate, and translateObject) where:
1. translateSubject may produce a set of triple patterns (to be considered as a union) 
2. then, for each of these triple patterns, translatePredicate is invoked, which also may produce a set of triple patterns for each input triple pattern---however, all of these triple patterns together (i.e., for all input triple patterns coming from translateSubject) can also still be considered as a single union
3. then the same happens again with translateObject

So, in the end, there is still only one set of triple patterns that can be considered as a union. Is that how it works or did I overlook something and things are more complicated?

Reply:
Something is wrong with the example. `owl:union` and `owl:intersectionOf` are not for properties but for classes, right?

Do you have any other examples where you would get a join (AND) of triple patterns?
Also, can it even be a join of something more complex than triple patterns (e.g., unions or other joins)?

Reply:
@JulietteVV No matter what your answers to my questions in my previous comment are, I see now that the return type of this function should not be `Set<TriplePattern>`. Even if it is only about making the meaning of this set more explicit, the return type should be an actual representation of a union of triple patterns (for the cases in which such a union is the intended result). Therefore, I propose that you change the return type to the general interface `SPARQLGraphPattern` (in `se.liu.ida.hefquin.engine.query`). Then, there are a number of relevant sub-interfaces that you can use for the actual results, depending on what the result is:
* If the result is just a single triple pattern, return a `TriplePattern` object.
* If the result is a join of triple patterns, return a `BGP` object.
* If the result is a union of triple patterns (or, more generally, a union of some graph patterns), return a `SPARQLUnionPattern` object.

The corresponding classes that implement these interfaces and that you can use can be found in the package `se.liu.ida.hefquin.engine.query.impl`.

Regarding the `SPARQLUnionPattern` interface, I have just added this interface (plus an implementation) to your branch for this purpose. Hence, you need to do a `git pull` to get it in your local branch.

Let me know if you have any questions.